### PR TITLE
AHOYAPPS-821: Change AudioDevice auto select order

### PR DESCRIPTION
**Pull request by Alton09**
## Description

Added a constructor parameter named `preferredDeviceList` to configure the order in which audio devices are automatically selected and activated when `selectedAudioDevice` is null.

## Breakdown
- Changed target sdk to 30
- Add Logger interface to allow unit test logging
- Made logs more readable
- Upgraded spotless plugin

## Validation
- Passed CI.
- Manually validated the automatic selection of each device type with a Pixel One XL and a Samsung Galaxy S10.

## Submission Checklist
 - [x] The source has been evaluated for semantic versioning changes and are reflected in `gradle.properties`
 - [x] The `CHANGELOG.md` reflects any **feature**, **bug fixes**, or **known issues** made in the source code

### CODE REVIEW

File: audioswitch/src/main/java/com/twilio/audioswitch/AudioSwitch.kt [hunk 1]
```diff
@@ -105,21 +103,36 @@ class AudioSwitch {
      *
      * @param context An Android Context.
      * @param loggingEnabled Toggle whether logging is enabled. This argument is false by default.
-     * @param audioFocusChangeListener A listener that is invoked when the system audio focus is
-     * updated. Note that updates are only sent to the listener after [activate] has been called.
+     * @param audioFocusChangeListener A listener that is invoked when the system audio focus is updated.
+     * Note that updates are only sent to the listener after [activate] has been called.
+     * @param preferredDeviceList The order in which [AudioSwitch] automatically selects and activates
+     * an [AudioDevice]. This parameter is ignored if the [selectedAudioDevice] is not null.
```
Comment (aaalaniz) [inline audioswitch/src/main/java/com/twilio/audioswitch/AudioSwitch.kt:109]:
```suggestion
     * an [AudioDevice]. This parameter is ignored if the [selectedAudioDevice] is not `null`.
```

File: audioswitch/src/main/java/com/twilio/audioswitch/AudioSwitch.kt [hunk 2]
```diff
@@ -105,21 +103,36 @@ class AudioSwitch {
...
+        preferredDeviceList: List<Class<out AudioDevice>> = defaultPreferredDeviceList
+    ) : this(context.applicationContext, ProductionLogger(loggingEnabled), audioFocusChangeListener,
+            preferredDeviceList)
 
     @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
     internal constructor(
         context: Context,
         logger: Logger,
         audioFocusChangeListener: OnAudioFocusChangeListener,
+        preferredDeviceList: List<Class<out AudioDevice>>,
```
Comment (aaalaniz) [inline audioswitch/src/main/java/com/twilio/audioswitch/AudioSwitch.kt:135]: The the `out AudioDevice` indicate that only instances of an `AudioDevice` class can be provided?

File: audioswitch/src/main/java/com/twilio/audioswitch/AudioSwitch.kt [hunk 3]
```diff
@@ -134,7 +147,24 @@ class AudioSwitch {
...
         this.wiredHeadsetReceiver = wiredHeadsetReceiver
         this.bluetoothHeadsetManager = headsetManager
+        this.preferredDeviceList = getPreferredDeviceList(preferredDeviceList)
         logger.d(TAG, "AudioSwitch($VERSION)")
+        logger.d(TAG, "Preferred device list = ${this.preferredDeviceList.map { it.simpleName }}")
+    }
+
+    private fun getPreferredDeviceList(preferredDeviceList: List<Class<out AudioDevice>>):
+            List<Class<out AudioDevice>> {
+        require(preferredDeviceList.isNotEmpty() && hasNoDuplicates(preferredDeviceList))
```
Comment (aaalaniz) [inline audioswitch/src/main/java/com/twilio/audioswitch/AudioSwitch.kt:157]: I think empty is okay. Providing an empty list results in the default preferred list being used.

File: audioswitch/src/main/java/com/twilio/audioswitch/AudioSwitch.kt [hunk 4]
```diff
@@ -105,21 +103,36 @@ class AudioSwitch {
...
      * @param context An Android Context.
      * @param loggingEnabled Toggle whether logging is enabled. This argument is false by default.
-     * @param audioFocusChangeListener A listener that is invoked when the system audio focus is
-     * updated. Note that updates are only sent to the listener after [activate] has been called.
+     * @param audioFocusChangeListener A listener that is invoked when the system audio focus is updated.
+     * Note that updates are only sent to the listener after [activate] has been called.
+     * @param preferredDeviceList The order in which [AudioSwitch] automatically selects and activates
+     * an [AudioDevice]. This parameter is ignored if the [selectedAudioDevice] is not null.
+     *
+     * The default preferred [AudioDevice] order is the following:
```
Comment (aaalaniz) [inline audioswitch/src/main/java/com/twilio/audioswitch/AudioSwitch.kt:111]: Add documentation that an `IllegalArgumentException` is thrown if the list contains duplicate entries.

File: audioswitch/src/main/java/com/twilio/audioswitch/AudioSwitch.kt (hunk 3 above)
Comment (aaalaniz) [reply audioswitch/src/main/java/com/twilio/audioswitch/AudioSwitch.kt:157]: Let's also ensure that an `IllegalArgumentException` is thrown when the list has duplicates.

**Review by aaalaniz — CHANGES_REQUESTED** [review]

File: audioswitch/src/main/java/com/twilio/audioswitch/AudioSwitch.kt (hunk 2 above)
Comment (Alton09) [reply audioswitch/src/main/java/com/twilio/audioswitch/AudioSwitch.kt:135]: Yep! It guarantees that only classes that extend AudioDevice are allowed.

File: audioswitch/src/main/java/com/twilio/audioswitch/AudioSwitch.kt (hunk 3 above)
Comment (Alton09) [reply audioswitch/src/main/java/com/twilio/audioswitch/AudioSwitch.kt:157]: Yep, I think that makes sense for an empty list. And an `IllegalArgumentException` does get thrown in the case of a duplicate.

**Review by aaalaniz — COMMENTED** [review]
Manual Test Cases

Data
1. Bluetooth, Wired, Speaker, Earpiece (Video)
2. Bluetooth, Wired, Earpiece, Speaker (Voice)

**Review by aaalaniz — APPROVED** [review]

**Pull request merged**